students/EricAdams/session06/test_mailroom.py

```diff
@@ -0,0 +1,221 @@
+#! /usr/bin/env python
+
+'''
+test_mailroom.py
+Excercise in testing mailroom.py using pytest
+'''
+
+import mailroom
+import pytest
+import sys
+from textwrap import dedent
+import os
+
+
+def test_goodbye_fail():
+    '''
+    Use an invalid value to get the test to fail
+    Expected result - fail
+    '''
+    result = mailroom.goodbye()
+    assert result == ('uit')
+
+
+def test_goodbye_pass():
+    '''
+    Provide a valid input, ('quit')
+    '''
+    result = mailroom.goodbye()
+    assert result == ('quit')
+
+
+def test_print_donors(capsys):
+    '''
+    This test captures stdout and asserts that there is an entry from
+    the donor_db in the screen output.
+    Saw this on stackoverflow.com.... seems to resolve the issue
+    of testing screen output.
+    '''
+    mailroom.print_donors()
+    out, err = capsys.readouterr()
+    sys.stdout.write(out)
+    donor = list(mailroom.donor_db.keys())
+    assert out.find(donor[0])
+
+
+def test_find_donor_key():
+    '''
+    Input a valid user from the mailroom.donor_db
+    '''
+    result = mailroom.find_donor('William Gates, III')
+    assert result == 'William Gates, III'
+
+
+def test_find_donor_nonexistant():
+    result = mailroom.find_donor('Eric Adams')
+    assert result is None
+
+
+def test_main_menu_selection_t():
+    '''
+    Pass the test by giving a user input of 't'
+    Simulate user input of 't'
+    :data - a text file
+            in the current directory, (test_main_menu_selection_t_inputs.txt),
+            is used as input data.
+            The value that the user would input is simulated
+            by making a one line entry into the file.
+            Example: for a user input of 't', enter the single
+            character 't' on the first line, (no quotes).
+
+    Saw this solution to the 'input' issue on stackoverflow.
+    Seems to work
+    '''
+    orig_stdin = sys.stdin
+    with open("test_main_menu_selection_t_inputs.txt") as sys.stdin:
+        result = mailroom.main_menu_selection()
+    sys.stdin = orig_stdin
+    assert result == "t"
+
+
+def test_gen_letter():
+    '''
+    Input a valid donor and assert that the letter is returned
+    '''
+    letter = dedent('''
+          Dear {}
+
+          Thank you for your very kind donation of ${:.2f}.
```

Comment:
careful here -- this works, and is fine for now, and is helpful in you'll catch it if the function totally fails.

But if anything changes about the letter format, this will then fail, and you'll need to update the test to match the code, which is kind of a pain.

So ideally you test what is critical about the result, rather than that it's exactly what it is at the moment.

For instance: did the donors name get properly put in the letter?
Did the donation amount get put in the letter, properly formatted (with the $)?

That sort of thing.